egui-wgpu/src/winit.rs

```diff
@@ -1,72 +1,202 @@
+
+    instance: Instance,
+    adapter: Option<Adapter>,
+    render_state: Option<RenderState>,
+    surface_state: Option<SurfaceState>,
+}
 
-        let egui_rpass = renderer::RenderPass::new(&device, surface_format, msaa_samples);
+impl<'a> Painter<'a> {
+    /// Manages [`wgpu`] state, including surface state, required to render egui.
+    ///
+    /// Only the [`wpu::Instance`] is initialized here. Device selection and the initialization
+    /// of render + surface state is deferred until the painter is given its first window target
+    /// via [`set_window`]. (Ensuring that a device that's compatible with the native window is
+    /// chosen)
+    ///
+    /// Before calling [`paint_and_update_texture`] a [`wgpu::Surface`] must be initialized
+    /// (and corresponding render state) by calling [`set_window`] once you have
+    /// a [`winit::window::Window`] with a valid `.native_window()` associated.
+    pub fn new(
+        backends: wgpu::Backends,
+        power_preference: wgpu::PowerPreference,
+        device_descriptor: wgpu::DeviceDescriptor<'a>,
+        present_mode: wgpu::PresentMode,
+        msaa_samples: u32,
+    ) -> Self {
+        let instance = wgpu::Instance::new(backends);
 
         Self {
+            power_preference,
+            device_descriptor,
+            present_mode,
+            msaa_samples,
+
+            instance,
+            adapter: None,
+            render_state: None,
+            surface_state: None,
+        }
+    }
+
+    async fn init_render_state(
+        &self,
+        adapter: &Adapter,
+        target_format: TextureFormat,
+    ) -> RenderState {
+        let (device, queue) =
+            pollster::block_on(adapter.request_device(&self.device_descriptor, None)).unwrap();
+
+        let egui_rpass = renderer::RenderPass::new(&device, target_format, self.msaa_samples);
+
+        RenderState {
             device,
             queue,
-            surface_config,
-            surface,
+            target_format,
             egui_rpass,
         }
     }
 
-    pub fn max_texture_side(&self) -> usize {
-        self.device.limits().max_texture_dimension_2d as usize
+    // We want to defer the initialization of our render state until we have a surface
+    // so we can take its format into account.
+    //
+    // After we've initialized our render state once though we expect all future surfaces
+    // will have the same format and so this render state will remain valid.
+    fn ensure_render_state_for_surface(&mut self, surface: &Surface) {
+        if self.adapter.is_none() {
+            let adapter =
+                pollster::block_on(self.instance.request_adapter(&wgpu::RequestAdapterOptions {
+                    power_preference: self.power_preference,
+                    compatible_surface: Some(surface),
+                    force_fallback_adapter: false,
+                }))
+                .unwrap();
+
+            self.adapter = Some(adapter);
+        }
```

Comment:
What I meant is this:

```suggestion
        self.adapter.get_or_insert_with(|| {
            pollster::block_on(self.instance.request_adapter(&wgpu::RequestAdapterOptions {
                power_preference: self.power_preference,
                compatible_surface: Some(surface),
                force_fallback_adapter: false,
            }))
            .unwrap()
        });
```

we don't care about what `get_or_insert_with` returns, so I don't see the borrow checker issue – but I might me missing something

Reply:
Yep, sorry I understood what you were suggesting, and it's ok to change the adapter initialization as suggesting but it doesn't work out to also change the render_state to be initialized in the same way which you had also suggested.

Actually I was mistaken in blaming the issue with the `render_state` initialization on the preceding adapter initialization. The problem with the render_state init is that the closure passed to `get_or_insert_with` also needs to borrow self when it calls `self.init_render_state`. (I was thinking the compiler was upset about the borrow returned from the preceding adapter initialization - sorry for the confusion)

So we could apply the change for the first block but then it ends up being inconsistent and then I personally didn't find that the code scanned better. Happy to make the suggested change still for initializing the adapter if you don't mind the inconsistency?

So just to be extra clear. If the code is changed to this:
```rust
    fn ensure_render_state_for_surface(&mut self, surface: &Surface) {
        self.adapter.get_or_insert_with(|| {
            pollster::block_on(self.instance.request_adapter(&wgpu::RequestAdapterOptions {
                power_preference: self.power_preference,
                compatible_surface: Some(surface),
                force_fallback_adapter: false,
            }))
            .unwrap()
        });

        self.render_state.get_or_insert_with(|| {
            let adapter = self.adapter.as_ref().unwrap();
            let swapchain_format = surface.get_preferred_format(adapter).unwrap();

            pollster::block_on(self.init_render_state(adapter, swapchain_format))
        });
    }
```
The compiler will complain like this:
```rust
error[E0502]: cannot borrow `self.render_state` as mutable because it is also borrowed as immutable
   --> C:\Users\Robert\src\egui\egui-wgpu\src\winit.rs:102:9
    |
102 |           self.render_state.get_or_insert_with(|| {
    |           ^                 ------------------ -- immutable borrow occurs here
    |           |                 |
    |  _________|                 immutable borrow later used by call
    | |
103 | |             let adapter = self.adapter.as_ref().unwrap();
104 | |             let swapchain_format = surface.get_preferred_format(adapter).unwrap();
105 | |
106 | |             let rs = pollster::block_on(self.init_render_state(adapter, swapchain_format));
    | |                                         ---- first borrow occurs due to use of `*self` in closure
107 | |             rs
108 | |         });
    | |__________^ mutable borrow occurs here
```

This works ok, even though it's inconsistent:
```rust
    fn ensure_render_state_for_surface(&mut self, surface: &Surface) {
        self.adapter.get_or_insert_with(|| {
            pollster::block_on(self.instance.request_adapter(&wgpu::RequestAdapterOptions {
                power_preference: self.power_preference,
                compatible_surface: Some(surface),
                force_fallback_adapter: false,
            }))
            .unwrap()
        });

        if self.render_state.is_none() {
            let adapter = self.adapter.as_ref().unwrap();
            let swapchain_format = surface.get_preferred_format(adapter).unwrap();

            let rs = pollster::block_on(self.init_render_state(adapter, swapchain_format));
            self.render_state = Some(rs);
        }
    }
```

Reply:
ah right, thanks for explaining!